engine/src/main/java/org/hibernate/search/spi/SearchIntegratorBuilder.java

```diff
@@ -403,6 +407,35 @@ else if ( metadataProvider.containsSearchMetadata( mappedClass ) ) {
 	}
 
 	/**
+	 * Get the filter definitions from the super classes that are not annotated with @Indexed
+	 */
+	private void bindFilterDefsInSuperClasses(final Map<XClass, Class<?>> classMappings) {
+		Set<XClass> superClasses = new HashSet<XClass>();
+		for ( Map.Entry<XClass, Class<?>> mapping : classMappings.entrySet() ) {
+			saveSuperClasses( superClasses, mapping.getKey() );
+		}
+		for ( XClass xClass : superClasses ) {
+			bindFilterDefs( xClass );
+		}
+	}
+
+	/**
+	 * It will add in the list all the super classes of a given class to the collection
+	 * if it's not annotated with @Indexed and it's not already in the collection.
+	 *
+	 * @param superClasses the list containing the super classes
+	 * @param mappedXClass the current class
+	 */
+	private void saveSuperClasses(Set<XClass> superClasses, XClass mappedXClass) {
+		if ( mappedXClass.getSuperclass() != null ) {
+			saveSuperClasses( superClasses, mappedXClass.getSuperclass() );
+		}
+		if ( !mappedXClass.isAnnotationPresent( Indexed.class ) && !OBJECT_CLASS_NAME.equals( mappedXClass.getName() ) ) {
```

Comment:
> why the condition of !mappedXClass.isAnnotationPresent( Indexed.class )? 

Initially, I didn't want to change the existing code and I kept the patch separated. The classes annotated with `@Indexed` are processed later so I decided to ignore them here. I will change this so that filters are processed in one part of the code only.

> That said, there is a ReflectionHelper.createXClassHierarchy( parseContext.getCurrentClass() ); which gives you the hierarchy as a list (bottom up). I think you could use that.

ok

> Last but not least, we are already processing the full hierarchy for indexed classes in AnnotationMetadataProvider. Now we do it twice as bootstrap. ... Seems to be a waste. Technically there would be no problem to find the filter definitions in AnnotationMetadataProvider instead and only process the class hierarchies once. 

I think I prefer the way it's done now, because the logic is going to be in one place. If I move it in the AnnotationMetadataProvider, I would be able to find the filters defined on one hierarchy but then I would have to check duplicates in `SearchIntegratorBuilder` anyway.
This way instead global elements are parsed in one place while type elements in another. I don't see it as waste unless there is reason to believe that this is going to become a performance bottleneck.